### DISCUSSION TITLE
Decoding uniswap v3 Router with 2 swaps

### DISCUSSION THREAD

**Discussion by Chief-Alchemist** [Elnaril/uniswap-universal-router-decoder#32] [Q&A] [answered]
Seems that the lib does not correctly decode for cases where there are 2 swaps in one, here is an example TX:
https://etherscan.io/tx/0x88d143d3b2cb85c2b12610f7d07f237728946ae68eb6c299e0ab31bd338744e0

The output seems pretty strange, and is definitely incorrect for the second leg, where USDT value is very large.

`(<Function execute(bytes,bytes[],uint256)>, {'commands': b'\x08\x00', 'inputs': [(<Function V2_SWAP_EXACT_IN(address,uint256,uint256,address[],bool)>, {'recipient': '0x0000000000000000000000000000000000000002', 'amountIn': 20991550410, 'amountOutMin': 0, 'path': ['0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', '0xdAC17F958D2ee523a2206206994597C13D831ec7'], 'payerIsSender': True}), (<Function V3_SWAP_EXACT_IN(address,uint256,uint256,bytes,bool)>, {'recipient': '0x0000000000000000000000000000000000000001', 'amountIn': 57896044618658097711785492504343953926634992332820282019728792003956564819968, 'amountOutMin': 8116820978936411681401, 'path': b'\xda\xc1\x7f\x95\x8d.\xe5#\xa2 b\x06\x99E\x97\xc1=\x83\x1e\xc7\x00\x0b\xb8aM\xa3\xb3{of\xf7\xcei\xb4\xbb\xbc\xf9\xa5\\\xe6\x16\x87\x07', 'payerIsSender': False})], 'deadline': 1699592747})`

**Comment by Elnaril** [comment]
Hello @Chief-Alchemist 
I'm closing this issue as duplicate of #10 and #14, but feel free to re-open it if you still believe there is an issue.
Thank you for your interest in this lib!

**Comment by Chief-Alchemist** [comment]
Thanks for the quick response, had a look, makes sense, one question I have left. You mention to solve for this we should:

> To handle that, you can set the amountIn to 2**255 which is the constant CONTRACT_BALANCE defined [here](https://github.com/Uniswap/universal-router/blob/a7bcf5459f475407a5af0aba6e5b6196e9461e4c/contracts/libraries/Constants.sol#LL11C8-L11C8).

How do we know whether we should or should not set amountIn? ie if there are more than 1 swaps overwrite the amountIn or?

**Comment by Elnaril** [comment, marked as answer]
(Converting issue #31 into a discussion)

When there is a first swap with an unknown `amountOut` (when the trx is encoded), then a second swap which has to use it as its own `amountIn`.

Worth to be noted, this SDK supports now 2 convenient methods that deal with this case for you:
- `v2_swap_exact_in_from_balance()`
- `v3_swap_exact_in_from_balance()`

**Reply by Elnaril** [reply]
@Chief-Alchemist , does it answer your question?

**Reply by Chief-Alchemist** [reply]
Yes, thanks!

**Comment by olvida** [comment]
Useful discussion, thanks! ;)

**Comment by KulzeinepHub** [comment]
Good!
